internal/controllers/stats.go

```diff
@@ -64,20 +63,90 @@ func (c *ApiController) StatisticsHandlerAllStatsForLoadbalancer(w http.Response
 		return
 	}
 
-	statistics, err := payout.GetStatsForPayout(c.repositories, getNow(), true)
+	totalRewardAsFloat, err := getTotalRewardFromRequest(r)
+	if err != nil {
+		log.Error(err)
+		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+	}
+
+	timestamp := getNow()
+	statistics, err := stats.CalculateStatisticsFromLastPayout(c.repositories, timestamp)
 	if err != nil {
 		log.Errorf("Failed to calculate statistics, because %v", err)
 		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
 		return
 	}
 
+	err = c.repositories.PayoutRepo.Save(&models.Payout{
+		Timestamp:      timestamp,
+		PaymentDetails: statistics,
+		LbFee:          totalRewardAsFloat * float64(configuration.Config.Fee),
+	})
+	if err != nil {
+		log.Errorf("Failed to save payout, because %v", err)
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	feesToNodes := payout.CalculatePayoutDistributionByNode(
+		statistics,
+		totalRewardAsFloat,
+		payout.LoadBalancerDistributionConfiguration{
+			FeePercentage:       float64(configuration.Config.Fee),
+			DifferentFeeAddress: false,
+		},
+	)
+	for nodeId, amount := range feesToNodes {
+		err := c.repositories.FeeRepo.RecordNewFee(nodeId, amount.Int64())
+		if err != nil {
+			log.Errorf("Failed to save fee for node %s, because %v", nodeId, err)
+		}
+	}
+
 	w.Header().Set("Content-Type", "application/json")
 	_ = json.NewEncoder(w).Encode(LoadbalancerStatsResponse{
 		Stats: statistics,
 		Fee:   configuration.Config.Fee,
 	})
 }
 
+func verifySignatureInHeader(r *http.Request, privateKey string) (bool, int, error) {
+	sig := r.Header.Get("X-Signature")
+	if sig == "" {
+		log.Error("Missing signature header")
+		return false, http.StatusBadRequest, nil
+	}
+	sigInBytes, err := hexutil.Decode(sig)
+	if err != nil {
+		log.Errorf("Unable to decode signature, because of: %v", err)
+		return false, http.StatusBadRequest, err
+	}
+	verified, err := signature.Verify([]byte(StatsSignedData), sigInBytes, privateKey)
+	if err != nil {
+		log.Errorf("Failed to verify signature, because %v", err)
+		return false, http.StatusInternalServerError, err
+	}
+	return verified, 0, err
+}
+
+func getTotalRewardFromRequest(r *http.Request) (float64, error) {
```

Comment:
Can this be in some service?

Reply:
I actually think this is not reusable code as it is tightly coupled with this specific handler so I would leave it here.